File: turn-one-client/lib/plan-features.ts
export type PlanType = "BASIC" | "PRO" | "ELITE";

export interface SimRacingPlanFeatures {
    multiChannelChart: boolean;
    comparison: boolean;
    replayScrubber: boolean;
    coachingTips: boolean;
    coachingChat: boolean;
    overlayTokens: boolean;
    spectatePublic: boolean;
    sessionLimit: number;
}

export const PLAN_FEATURES: Record<PlanType, SimRacingPlanFeatures> = {
    BASIC: {
        multiChannelChart: false,
        comparison: false,
        replayScrubber: false,
        coachingTips: false,
        coachingChat: false,
        overlayTokens: false,
        spectatePublic: false,
        sessionLimit: 1,
    },
    PRO: {
        multiChannelChart: true,
        comparison: true,
        replayScrubber: true,
        coachingTips: true,
        coachingChat: false,
        overlayTokens: true,
        spectatePublic: true,
        sessionLimit: 5,
    },
    ELITE: {
        multiChannelChart: true,
        comparison: true,
        replayScrubber: true,
        coachingTips: true,
        coachingChat: true,
        overlayTokens: true,
        spectatePublic: true,
        sessionLimit: 15,
    },
};

export function planFeatures(plan: PlanType | string | null | undefined): SimRacingPlanFeatures {
    if (plan && plan in PLAN_FEATURES) return PLAN_FEATURES[plan as PlanType];
    return PLAN_FEATURES.BASIC;
}

File: turn-one-backend/Domain/PlanFeatures.cs
using Domain.Enums;

namespace Domain;

public record SimRacingPlanFeatures(
    bool MultiChannelChart,
    bool Comparison,
    bool ReplayScrubber,
    bool CoachingTips,
    bool CoachingChat,
    bool OverlayTokens,
    bool SpectatePublic,
    int SessionLimit
);

public static class PlanFeatures
{
    public static SimRacingPlanFeatures SimRacing(PlanType plan) => plan switch
    {
        PlanType.BASIC => new(
            MultiChannelChart: false,
            Comparison: false,
            ReplayScrubber: false,
            CoachingTips: false,
            CoachingChat: false,
            OverlayTokens: false,
            SpectatePublic: false,
            SessionLimit: 1
        ),
        PlanType.PRO => new(
            MultiChannelChart: true,
            Comparison: true,
            ReplayScrubber: true,
            CoachingTips: true,
            CoachingChat: false,
            OverlayTokens: true,
            SpectatePublic: true,
            SessionLimit: 5
        ),
        PlanType.ELITE => new(
            MultiChannelChart: true,
            Comparison: true,
            ReplayScrubber: true,
            CoachingTips: true,
            CoachingChat: true,
            OverlayTokens: true,
            SpectatePublic: true,
            SessionLimit: 15
        ),
        _ => SimRacing(PlanType.BASIC)
    };
}
